featurebyte/service/feature_table_cache.py

```diff
@@ -1,111 +1,448 @@
+        db_session: BaseSession,
+        graph: QueryGraph,
+        non_cached_nodes: List[Tuple[Node, CachedFeatureDefinition]],
+        serving_names_mapping: Optional[Dict[str, str]] = None,
+        is_feature_list_deployed: bool = False,
+    ) -> None:
+        # create temporary table with features
+        temp_final_cache_table_name = (
+            f"__TEMP__{MaterializedTableNamePrefix.FEATURE_TABLE_CACHE}_{ObjectId()}"
+        )
+
+        merge_target_table_alias = "feature_table_cache"
+        merge_source_table_alias = "partial_features"
+
+        try:
+            await self._create_table(
+                feature_store=feature_store,
+                observation_table=observation_table,
+                db_session=db_session,
+                final_table_name=temp_final_cache_table_name,
+                graph=graph,
+                nodes=non_cached_nodes,
+                serving_names_mapping=serving_names_mapping,
+                is_feature_list_deployed=is_feature_list_deployed,
+            )
+
+            # alter cached tables adding columns for new features
+            adapter = get_sql_adapter(db_session.source_type)
+            table_exr = expressions.Table(
+                this=quoted_identifier(cache_metadata.table_name),
+                db=quoted_identifier(db_session.schema_name),
+                catalog=quoted_identifier(db_session.database_name),
+            )
+            columns_expr = [
+                expressions.ColumnDef(
+                    this=quoted_identifier(definition.feature_name),
+                    kind=expressions.DataType.build(
+                        adapter.get_physical_type_from_dtype(extract_dtype_from_graph(graph, node))
+                    ),
+                )
+                for node, definition in non_cached_nodes
+            ]
+            alter_table_sql = alter_table_add_columns_sql(
+                table_exr, columns_expr, db_session.source_type
+            )
+            await db_session.execute_query(alter_table_sql)
+
+            # merge temp table into cache table
+            merge_conditions = [
+                expressions.EQ(
+                    this=get_qualified_column_identifier(
+                        InternalName.TABLE_ROW_INDEX, merge_target_table_alias
+                    ),
+                    expression=get_qualified_column_identifier(
+                        InternalName.TABLE_ROW_INDEX, merge_source_table_alias
+                    ),
+                )
+            ]
+            update_expr = expressions.Update(
+                expressions=[
+                    expressions.EQ(
+                        this=get_qualified_column_identifier(
+                            definition.feature_name, merge_target_table_alias
+                        ),
+                        expression=get_qualified_column_identifier(
+                            definition.feature_name, merge_source_table_alias
+                        ),
```

Comment:
Currently `_update_table()` will create two tables: one created by `get_historical_features`, one created by renaming the original feature names to cache table column names. Both are dropped in the end.

I think it can be improved by only creating only one table: don't create the second table for renaming purpose, and instead use the first table for the merge operation. Update `update_expr` here to account for that (specify the original feature name before renaming).

Not sure the impact, but maybe something we can revisit later.